### PR TITLE
Added constructor details in error message

### LINKED ISSUE #25: Error message details from the container
[Errors](https://github.com/uber-go/dig/blob/master/container.go#L33-L37) defined in container right now are not useful in mentioning which constructor failed the execution. 

example, A useful error would be - 
"constructor return type must be a pointer, constructor: `func foo() *ReturnType`"

### PR DESCRIPTION
**Pull request by anuptalwalkar**
Fixes #25 
error message now has details of the constructor and function signature.

### CODE REVIEW

File: container.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (c *Container) Invoke(t interface{}) error {
 	case reflect.Func:
 		args, err := c.Graph.ConstructorArguments(ctype)
 		if err != nil {
-			return err
+			return errors.Wrapf(err, "error invoking function %v", t)
```
Comment (alsamylkin) [inline container.go]: Is it going to print actual function, not an address?

File: container_test.go [hunk 2]
```diff
@@ -72,7 +72,7 @@ func TestRegister(t *testing.T) {
 			err := c.Provide(tc.param)
 
 			if tc.err != nil {
-				require.EqualError(t, err, tc.err.Error())
+				require.Contains(t, err.Error(), tc.err.Error())
```
Comment (alsamylkin) [inline container_test.go]: Can we add a test that check, that function name is present in the error message?

File: container.go (hunk 1 above)
Comment (anuptalwalkar) [inline container.go]: yes. its printing `ctype`, which is actual func signature

File: container_test.go (hunk 2 above)
Comment (anuptalwalkar) [inline container_test.go]: Will do.

File: container_test.go [hunk 3]
```diff
@@ -57,22 +57,22 @@ func TestRegister(t *testing.T) {
...
+		err   string
 	}{
-		{"non function", struct{}{}, errParamType},
-		{"wrong return count", noReturn, errReturnCount},
-		{"non pointer return", returnNonPointer, errReturnKind},
-		{"wrong parameters type", nonPointerParams, errArgKind},
-		{"pointer Provide", &struct{}{}, nil},
+		{"non function", struct{}{}, "for constructor struct {}: " + errParamType.Error()},
+		{"wrong return count", noReturn, "for constructor func(): " + errReturnCount.Error()},
+		{"non pointer return", returnNonPointer, "for constructor func() dig.S: " + errReturnKind.Error()},
```
Comment (alsamylkin) [inline container_test.go]: where is `returnNonPointer` in the error message?

Comment (anuptalwalkar) [inline container_test.go]: no, that is not the intention. The intention was to identify what func call is the cause. e.g., knowing "func() dig.S" returns no pointer is more valuable.

Comment (alsamylkin) [inline container_test.go]: We can do better:
```golang
runtime.FuncForPC(reflect.ValueOf(functionOnj).Pointer()).Name()
```

File: container.go [hunk 4]
```diff
@@ -117,7 +117,7 @@ func (c *Container) Provide(types ...interface{}) error {
 				return err
 			}
 		default:
-			return errParamType
+			return errors.Wrapf(errParamType, "for constructor %v", ctype)
```
Comment (glibsm) [inline container.go]: "for constructor %v" is used a lot, can we const-ify that?

File: container.go [hunk 5]
```diff
@@ -76,11 +76,11 @@ func (c *Container) Invoke(t interface{}) error {
 			case reflect.Slice, reflect.Array, reflect.Map, reflect.Ptr, reflect.Interface:
 				c.Graph.InsertObject(values[i])
 			default:
-				return errors.Wrapf(errReturnKind, "%v", ctype)
+				return errors.Wrapf(errReturnKind, "error invoking function %v", ctype)
 			}
 		}
 	default:
-		return errParamType
+		return errors.Wrapf(errParamType, "error invoking function %v", ctype)
```
Comment (glibsm) [inline container.go]: same question here

**Pull request merged**